File: ios/App/Plugins/Phrase/ScoringConfig.swift
import Foundation

/// Configuração centralizada de pesos, thresholds, cooldowns e limites do pipeline.
/// Paridade com ScoringConfig.java (Android).
class ScoringConfig {
    // MARK: - Fatores acústicos
    var acousticRmsWeight: Double = 0.3
    var acousticSpeechBurstWeight: Double = 0.4
    var acousticDurationWeight: Double = 0.3

    // MARK: - Fatores semânticos (frases detectadas)
    var semanticGenericWeight: Double = 0.2
    var semanticPersonalizedWeight: Double = 0.5
    var semanticOperationalWeight: Double = 0.0  // Operacionais não somam score

    // MARK: - Penalizações
    var stablePatternPenalty: Double = -0.3
    var stablePatternRmsVarianceThreshold: Double = 3.0
    var stablePatternWindowSeconds: Int = 30

    // MARK: - Thresholds de risco
    var riskLowMax: Double = 0.3
    var riskMediumMax: Double = 0.6
    var riskHighMin: Double = 0.6

    // MARK: - DTW matching
    var dtwDefaultThreshold: Double = 80.0
    var dtwConfidenceMin: Double = 0.5
    var dtwAbandonFactor: Double = 1.5

    // MARK: - Limites de frases (SPEC v3 §9.4)
    var maxOperationalPhrases: Int = 4
    var maxContextualPhrases: Int = 10
    var maxGenericPhrases: Int = 8
    var minSamplesPerPhrase: Int = 3
    var maxSamplesPerPhrase: Int = 5

    // MARK: - Enrollment
    var enrollmentConsistencyMaxDistance: Double = 120.0
    var enrollmentConflictMinDistance: Double = 2.0
    var enrollmentThresholdFactor: Double = 4.0

    // MARK: - Segurança por comando (SPEC v3 §10)
    struct CommandSecurity {
        let confidenceMin: Double
        let repetitionsRequired: Int
        let cooldownSeconds: Int
    }

    var commandSecurity: [String: CommandSecurity] = [
        "start_recording": CommandSecurity(confidenceMin: 0.55, repetitionsRequired: 1, cooldownSeconds: 5),
        "stop_recording":  CommandSecurity(confidenceMin: 0.65, repetitionsRequired: 1, cooldownSeconds: 10),
        "start_panic":     CommandSecurity(confidenceMin: 0.85, repetitionsRequired: 2, cooldownSeconds: 30),
        "cancel_panic":    CommandSecurity(confidenceMin: 0.9, repetitionsRequired: 3, cooldownSeconds: 60),
    ]

    // MARK: - Repetição
    var repetitionBufferMaxAge: TimeInterval = 15.0
    var repetitionBufferMaxEntries: Int = 8

    // MARK: - Confirmation window
    var confirmationWindowPreSeconds: Double = 3.0
    var confirmationWindowPostSeconds: Double = 3.0

    // MARK: - Mode B (escuta contínua leve)
    var modeBEnabled: Bool = true
    var modeBIntervalMs: Int = 500
    var modeBMaxPhrases: Int = 4
}

File: android/app/src/main/java/tech/orizon/ampara/audio/phrase/ScoringConfig.java
package tech.orizon.ampara.audio.phrase;

import java.util.HashMap;
import java.util.Map;

/**
 * Configuração centralizada de scoring e thresholds do pipeline.
 * 
 * TODOS os pesos, limiares e limites ficam aqui.
 * Nenhum hardcode disperso pelo código.
 * Conforme SPEC v3, seção 10.
 */
public class ScoringConfig {

    // ========== Pesos acústicos ==========
    public double weightSpeechDetected = 0.5;
    public double weightAnomalousSequence = 1.5;
    public double weightAbruptChange = 1.0;
    // weightVoiceOverlap: REMOVIDO da v1 (candidato v2/v3)
    public double weightImpactEvent = 2.5;
    public double weightMultipleEvents = 1.5;

    // ========== Padrão estável (TV/música) — heurística v1 ==========
    // Detecção via variância do RMS em janela de 30s
    public double stablePatternVarianceThreshold = 3.0; // dB²
    public int stablePatternWindowSeconds = 30;

    // ========== Pesos semânticos genéricos (APOIO CONTEXTUAL APENAS) ==========
    public Map<String, Double> genericKeywordWeights;
    /** Vocabulário genérico NUNCA dispara ações críticas sozinho */
    public boolean genericCanTriggerAction = false;

    // ========== Pesos de frases personalizadas ==========
    public double weightOperationalHigh = 8.0;
    public double weightContextualAlert = 3.0;
    public double repetitionBonus = 2.0;
    public double acousticSemanticComboBonus = 1.5;

    // ========== Penalizações ==========
    public double penaltyIsolatedEvent = -1.0;
    public double penaltyMarginalNoPhrase = -1.0;
    /** Aplicada SOMENTE se variância < stablePatternVarianceThreshold */
    public double penaltyStablePattern = -1.5;

    // ========== Faixas de score ==========
    public double bandLow = 2.9;
    public double bandMedium = 6.9;
    public double bandHigh = 10.9;
    // 11+ = crítico

    // ========== Segurança por comando (SPEC seção 6) ==========
    public double startRecordingConfidence = 0.55;
    public double stopRecordingConfidence = 0.65;
    /** Gravação não pode ser parada nos primeiros N ms */
    public int stopRecordingMinDurationMs = 30000;
    public double startPanicConfidence = 0.85;
    /** Se confidence >= instantConfidence, dispara sem exigir repetição */
    public double startPanicInstantConfidence = 0.92;
    public int startPanicRepetitionWindowMs = 15000;
    public double cancelPanicConfidence = 0.90;
    /** Cancelar pânico SEMPRE exige repetição */
    public int cancelPanicRepetitions = 2;
    public int cancelPanicRepetitionWindowMs = 20000;

    // ========== Cooldowns por comando ==========
    public int startRecordingCooldownMs = 5000;
    public int stopRecordingCooldownMs = 10000;
    public int startPanicCooldownMs = 30000;
    public int cancelPanicCooldownMs = 60000;

    // ========== Modo B (escuta contínua) ==========
    /** Boost nos thresholds quando em escuta contínua (evita falso positivo) */
    public double continuousModeConfidenceBoost = 0.10;

    // ========== Limites de frases (SPEC seção 9.4) ==========
    public int maxOperationalPhrases = 4;
    public int maxContextualPhrases = 10;
    public int minSamplesPerPhrase = 3;
    public int maxSamplesPerPhrase = 5;

    // ========== Histórico de repetição (SPEC seção 10.2) ==========
    /** Limpa entradas mais velhas que N ms */
    public int repetitionHistoryMaxAgeMs = 15000;
    public int repetitionHistoryMaxEntries = 8;

    // ========== Confirmation window (L2) ==========
    public int preTriggerSeconds = 3;
    public int postTriggerSeconds = 3;

    // ========== Enrollment ==========
    /** Duração mínima de frase para aceitar (ms) */
    public int enrollMinDurationMs = 500;
    /** Duração máxima de frase para aceitar (ms) */
    public int enrollMaxDurationMs = 5000;
    /** DTW máxima intra-amostras (consistência - AUMENTADO PARA NÃO FALHAR TANTO) */
    public double enrollConsistencyThreshold = 120.0;
    /** DTW mínima inter-frases operacionais (conflito - DIMINUIDO AO EXTREMO PARA PERMITIR 'INICIAR GRAVACAO' E 'PARAR GRAVACAO') */
    public double enrollDivergenceThreshold = 2.0;

    /**
     * Construtor com valores padrão e vocabulário genérico
     */
    public ScoringConfig() {
        genericKeywordWeights = new HashMap<>();
        genericKeywordWeights.put("nao", 1.0);
        genericKeywordWeights.put("para", 2.0);
        genericKeywordWeights.put("chega", 1.5);
        genericKeywordWeights.put("me_deixa", 2.5);
        genericKeywordWeights.put("ajuda", 3.5);
        genericKeywordWeights.put("policia", 3.0);
        genericKeywordWeights.put("socorro", 4.5);
        genericKeywordWeights.put("me_solta", 4.5);
    }

    /**
     * Retorna a confiança mínima para um tipo de comando
     */
    public double getMinConfidence(String commandId) {
        switch (commandId) {
            case "start_recording":
                return startRecordingConfidence;
            case "stop_recording":
                return stopRecordingConfidence;
            case "start_panic":
                return startPanicConfidence;
            case "cancel_panic":
                return cancelPanicConfidence;
            default:
                return 0.75; // padrão para contextuais
        }
    }

    /**
     * Retorna o cooldown para um tipo de comando (ms)
     */
    public int getCooldownMs(String commandId) {
        switch (commandId) {
            case "start_recording":
                return startRecordingCooldownMs;
            case "stop_recording":
                return stopRecordingCooldownMs;
            case "start_panic":
                return startPanicCooldownMs;
            case "cancel_panic":
                return cancelPanicCooldownMs;
            default:
                return 5000;
        }
    }

    /**
     * Classifica nível de risco com base no score
     */
    public String classifyRiskLevel(double score) {
        if (score > bandHigh)
            return "CRITICAL";
        if (score > bandMedium)
            return "HIGH";
        if (score > bandLow)
            return "MEDIUM";
        return "LOW";
    }
}
